packages/manager/src/features/Volumes/utils.ts

```diff
@@ -0,0 +1,78 @@
+import type { Event, Notification, Volume } from '@linode/api-v4';
+import type { Status } from 'src/components/StatusIcon/StatusIcon';
+
+export const volumeStatusIconMap: Record<Volume['status'], Status> = {
+  active: 'active',
+  creating: 'other',
+  migrating: 'other',
+  offline: 'inactive',
+  resizing: 'other',
+};
+
+/**
+ * Given an in-progress event and a volume's status, this function
+ * returns a volume's status with event info taken into account.
+ *
+ * We do this to provide users with a real-time feeling experience
+ * without having to refetch a volume's status agressivly.
+ *
+ * @param status The actual volume status from the volumes endpoint
+ * @param event An in-progress event for the volume
+ * @returns a volume status
+ */
+export const getDerivedVolumeStatusFromStatusAndEvent = (
+  status: Volume['status'],
+  event: Event | undefined
+): Volume['status'] => {
+  if (event === undefined) {
+    return status;
+  }
+
+  if (event.action === 'volume_migrate' && event.status === 'started') {
+    return 'migrating';
+  }
+
+  return status;
+};
+
+/**
+ * Returns a nicely formated percentage from an event
+ * only if the event is in progress and has a percentage.
+ *
+ * This allows us to show the user the progress of a
+ * volume migration.
+ *
+ * @returns "(50%)" for example
+ */
+export const getEventProgress = (event: Event | undefined) => {
+  if (
+    event === undefined ||
+    event.percent_complete === null ||
+    event.status !== 'started'
+  ) {
+    return null;
+  }
+
+  return `(${event.percent_complete}%)`;
+};
+
+export const getUpgradeableVolumeIds = (
+  volumes: Volume[],
+  notifications: Notification[]
+) => {
+  const upgradeableVolumeIds: number[] = [];
+
+  for (const volume of volumes) {
+    if (
+      notifications?.some(
+        (notification) =>
+          notification.entity?.id === volume.id &&
+          notification.type === 'volume_migration_scheduled'
+      )
+    ) {
+      upgradeableVolumeIds.push(volume.id);
+    }
+  }
+
+  return upgradeableVolumeIds;
```

Comment:
Definite nitpick, but I think this could be optimized by filtering out any non-`volume_migration_scheduled` notifications before iterating:

```suggestion
  return notifications
    .filter((notification) => notification.type === 'volume_migration_scheduled')
    .reduce((volumeIds: number[], notification: Notification) => {
      const upgradeableVolume = volumes.find((volume) => volume.id === notification.entity?.id);
      if (upgradeableVolume) {
        volumeIds.push(upgradeableVolume.id);
      }
      return volumeIds;
    }, []);
```

I did a quick experiment by adapting your unit test to call the function 100,000 times against arrays of 100 volumes and 250 notifications, and got ~3.5s with the function above vs ~8s with the original implementation. (I did this purely out of curiosity -- I think there's exactly zero chance this function ever becomes a bottleneck, but on the other hand it felt silly to scrap this, so just thought I'd share 🤷‍♂️)

Reply:
Great call. 

I didn't stress too much about perf because I believe Linodes can only have 8 volumes max, but this is much better.